File: retrocpu_test_framework/src/tms9995/cdb.ts
/**
 * TMS9995 向け CDB パーサ。
 * MN1613 の `parseCdb` は奇数バイトアドレスを拒否するため、バイト CPU では本関数を使う。
 * レコード形式は SDCC CDB（L:）と同じ。
 */

import type { CdbCheckpoint, CdbSymbol } from "../../../retrocpu_emu/src/code_test/types.js";
import type { CdbTable } from "../../../retrocpu_emu/src/code_test/cdb.js";

/** `L:__CP$uart_initialized$0001` の `$` 区切り本体 */
const CP_LEFT = /^__CP\$([A-Za-z_][A-Za-z0-9_]*)\$([0-9]{4})$/;

/**
 * 空の CDB 表を作る。
 * @returns シンボル／チェックポイントなし
 */
export function emptyTms9995CdbTable(): CdbTable {
  return { byName: new Map(), symbols: [], checkpoints: [] };
}

/**
 * CDB テキストをパースする（奇数バイトアドレス可）。
 * @param cdbText CDB 全文
 * @returns ラベル表とチェックポイント
 */
export function parseTms9995Cdb(cdbText: string): CdbTable {
  const byName = new Map<string, CdbSymbol>();
  const symbols: CdbSymbol[] = [];
  const checkpoints: CdbCheckpoint[] = [];

  const lines = cdbText.replace(/\r\n/g, "\n").split("\n");
  for (const raw of lines) {
    const line = raw.trim();
    if (!line || line[1] !== ":") continue;
    if (line[0] !== "L") continue;

    const body = line.slice(2);
    const colon = body.lastIndexOf(":");
    if (colon < 0) continue;
    const left = body.slice(0, colon);
    const addrHex = body.slice(colon + 1).trim();
    if (!/^[0-9A-Fa-f]+$/.test(addrHex)) continue;

    const byteAddr = parseInt(addrHex, 16) >>> 0;

    const cpLeft = left.startsWith("G$") ? left.slice(2) : left;
    if (cpLeft.startsWith("__CP$")) {
      const cpMatch = cpLeft.match(CP_LEFT);
      if (!cpMatch) {
        throw new Error(`Invalid checkpoint CDB record "${left}"`);
      }
      const name = cpMatch[1]!;
      const serial = cpMatch[2]!;
      checkpoints.push({
        id: `__CP$${name}$${serial}`,
        name,
        serial,
        byteAddr,
        wordAddr: byteAddr >>> 1,
      });
      continue;
    }

    const parts = left.split("$");
    if (parts.length < 2) continue;
    const scope = parts[0]!;
    const namePart = parts[1]!;
    if (!namePart) continue;

    const sym: CdbSymbol = {
      name: namePart,
      byteAddr,
      wordAddr: byteAddr >>> 1,
      scope,
    };
    symbols.push(sym);
    const prev = byName.get(namePart);
    if (!prev || scope === "G" || prev.scope !== "G") {
      byName.set(namePart, sym);
    }
  }

  return { byName, symbols, checkpoints };
}

/**
 * シンボルを名前で引く（大文字小文字無視）。
 * @param table parseTms9995Cdb の結果
 * @param name ラベル名
 * @returns 見つかったシンボル
 * @throws 未登録の場合
 */
export function requireTms9995Symbol(
  table: CdbTable,
  name: string,
): CdbSymbol {
  const exact = table.byName.get(name);
  if (exact) return exact;
  const upper = name.toUpperCase();
  for (const [k, v] of table.byName) {
    if (k.toUpperCase() === upper) return v;
  }
  throw new Error(`CDB symbol not found: ${name}`);
}

File: retrocpu_test_framework_rs/src/tms9995/cdb.rs
use std::collections::HashMap;

use crate::error::FrameworkError;

use super::types::{CdbCheckpoint, CdbSymbol, CdbTable};

fn is_hex(s: &str) -> bool {
    !s.is_empty() && s.bytes().all(|b| b.is_ascii_hexdigit())
}

pub fn empty_tms9995_cdb_table() -> CdbTable {
    CdbTable {
        by_name: HashMap::new(),
        symbols: Vec::new(),
        checkpoints: Vec::new(),
    }
}

pub fn parse_tms9995_cdb(cdb_text: &str) -> Result<CdbTable, FrameworkError> {
    let mut table = empty_tms9995_cdb_table();

    for raw in cdb_text.replace("\r\n", "\n").split('\n') {
        let line = raw.trim();
        if line.len() < 3 {
            continue;
        }
        if !line.starts_with("L:") {
            continue;
        }

        let body = &line[2..];
        let Some(colon) = body.rfind(':') else {
            continue;
        };
        let left = &body[..colon];
        let addr_hex = body[colon + 1..].trim();
        if !is_hex(addr_hex) {
            continue;
        }

        let byte_addr = u32::from_str_radix(addr_hex, 16).map_err(|_| {
            FrameworkError::invalid_argument(format!("invalid CDB address: {addr_hex}"))
        })?;

        let cp_left = left.strip_prefix("G$").unwrap_or(left);
        if let Some(rest) = cp_left.strip_prefix("__CP$") {
            let mut parts = rest.split('$');
            let Some(name) = parts.next() else {
                return Err(FrameworkError::invalid_argument(format!(
                    "invalid checkpoint CDB record \"{left}\""
                )));
            };
            let Some(serial) = parts.next() else {
                return Err(FrameworkError::invalid_argument(format!(
                    "invalid checkpoint CDB record \"{left}\""
                )));
            };
            if parts.next().is_some()
                || name.is_empty()
                || !name
                    .bytes()
                    .all(|b| b == b'_' || b.is_ascii_alphanumeric())
                || !name
                    .bytes()
                    .next()
                    .map(|b| b == b'_' || b.is_ascii_alphabetic())
                    .unwrap_or(false)
                || serial.len() != 4
                || !serial.bytes().all(|b| b.is_ascii_digit())
            {
                return Err(FrameworkError::invalid_argument(format!(
                    "invalid checkpoint CDB record \"{left}\""
                )));
            }

            table.checkpoints.push(CdbCheckpoint {
                id: format!("__CP${name}${serial}"),
                name: name.to_string(),
                serial: serial.to_string(),
                byte_addr,
                word_addr: byte_addr >> 1,
            });
            continue;
        }

        let mut parts = left.split('$');
        let Some(scope) = parts.next() else {
            continue;
        };
        let Some(name_part) = parts.next() else {
            continue;
        };
        if name_part.is_empty() {
            continue;
        }

        let sym = CdbSymbol {
            name: name_part.to_string(),
            byte_addr,
            word_addr: byte_addr >> 1,
            scope: scope.to_string(),
        };
        table.symbols.push(sym.clone());

        let prev = table.by_name.get(name_part);
        if prev.is_none() || scope == "G" || prev.map(|p| p.scope.as_str()) != Some("G") {
            table.by_name.insert(name_part.to_string(), sym);
        }
    }

    Ok(table)
}

pub fn require_tms9995_symbol(table: &CdbTable, name: &str) -> Result<CdbSymbol, FrameworkError> {
    if let Some(sym) = table.by_name.get(name) {
        return Ok(sym.clone());
    }

    let upper = name.to_ascii_uppercase();
    for (k, v) in &table.by_name {
        if k.to_ascii_uppercase() == upper {
            return Ok(v.clone());
        }
    }

    Err(FrameworkError::invalid_argument(format!(
        "CDB symbol not found: {name}"
    )))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_accepts_odd_byte_address() {
        let t = parse_tms9995_cdb("L:G$ODD$0$0:0001\n").expect("parse should work");
        assert_eq!(t.by_name.get("ODD").map(|s| s.byte_addr), Some(1));
    }

    #[test]
    fn parse_checkpoint_record() {
        let t = parse_tms9995_cdb("L:G$__CP$enter$0001:0101\n").expect("parse should work");
        assert_eq!(t.checkpoints.len(), 1);
        assert_eq!(t.checkpoints[0].id, "__CP$enter$0001");
        assert_eq!(t.checkpoints[0].byte_addr, 0x0101);
    }
}
